### ISSUE TITLE
PG::UnableToSend - number of parameters must be between 0 and 65535

### ISSUE THREAD

**Issue by packerbacker89** [rails/rails#35251]
### Steps to reproduce
Prepare the database with 65536 rows: `65536.times { Car.create }`

Place the following line into the top of a controller action: `Car.where(id: Car.all.ids).pluck(:id)`

Go to the route for that controller action

### Expected behavior
The ids of the cars should be returned (nothing would be seen in the browser, but the query would be seen in the logs)

### Actual behavior
An exception is thrown:
```
Completed 500 Internal Server Error
PG::UnableToSend - number of parameters must be between 0 and 65535
```

### System configuration
**Rails version**: 5.2.2

**Ruby version**: 2.5.3

This was partially addressed with this issue: https://github.com/rails/rails/issues/33702, which led to 5.2.2 having solved the problem when running queries in the rails console. However, the issue still exists for queries in the controllers (and models?)

**Comment by rafaelfranca** [comment]
> which led to 5.2.2 having solved the problem when running queries in the rails console. However, the issue still exists for queries in the controllers (and models?)

I don't see how this could happen. The code is not conditional to if a query is executed in a console or in a controller. It should behaves always in the same way. 

Could you create a reproduction script with this template https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_gem.rb?

**Comment by packerbacker89** [comment]
I agree that the different behavior in console / server is odd.

I've created a dummy app that replicates the issue: https://github.com/packerbacker89/pg_parameter_64k_issue

I was unable to replicate the issue, however, using the template you provided.

**Comment by matthewdunbar** [comment]
I'll see if I can replicate the issue using the template, but I am also seeing this behavior. Specifically, I am using `activerecord-sqlserver-adapter` version 5.2.0.rc1, Rails version 5.2.2, and Ruby version 2.5.1. Everything runs correctly in my rspec tests, but not in the full system tests. The most that I've been able to figure out is the parameterized SQL query and binds are being passed to the database adapter.

**Comment by ardavis** [comment]
I cloned @packerbacker89's repo and can confirm that it acts differently in the rails console versus in a controller.

```
$ rails c
2.5.3 :001 > @post_count = Post.where(id: Post.all.ids).pluck(:id)
   (16.5ms)  SELECT "posts"."id" FROM "posts"
   (307.9ms)  SELECT "posts"."id" FROM "posts" WHERE "posts"."id" IN (1, 2, 3, ...)
 => [1, 2, 3, 4, 5, ...]
```

The console works great, notice how the database statement was created.

Here's the output when you run a rails server and view the index page (after running the `/setup` to generate 70,000 Posts)

```
Started GET "/" for 127.0.0.1 at 2019-02-14 14:17:01 -0500
   (0.6ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
  ↳ /path/to/ruby-2.5.3/gems/activerecord-5.2.2/lib/active_record/log_subscriber.rb:98
Processing by PostsController#index as HTML
   (18.4ms)  SELECT "posts"."id" FROM "posts"
  ↳ app/controllers/posts_controller.rb:4
   (7.6ms)  SELECT "posts"."id" FROM "posts" WHERE "posts"."id" IN ($1, $2, $3, ...) [["id", 1], ["id", 2], ["id", 3], ...]
  ↳ app/controllers/posts_controller.rb:4
Completed 500 Internal Server Error in 717ms (ActiveRecord: 30.2ms)

ActiveRecord::StatementInvalid (PG::UnableToSend: number of parameters must be between 0 and 65535
```

It's very strange how the way it generates the database statement differs when calling it in the rails server versus the console.

**Comment by matthewdunbar** [comment]
Here's a script to reproduce this issue. It has something to do with the query cache being enabled.
```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "5.2.2"
  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "postgresql", host: '127.0.0.1', database: "postgres", password: 'postgres', username: 'postgres')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class BugTest < Minitest::Test
  def test_query_parameters
    Post.connection.enable_query_cache!

    Post.where(id: (1..65536).to_a).count
  end
end
```

**Comment by matthewdunbar** [comment]
My first guess is its related to this conditional: https://github.com/rails/rails/blob/6183483fa13bb1cac7e32f64df568fc2f9fb9a6b/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb#L96

**Comment by matthewdunbar** [comment]
I have opened a PR with a potential fix for this issue: https://github.com/rails/rails/pull/35286

**Comment by kamipo** [comment]
Fixed by #35286.

**Comment by crevete** [comment]
@kamipo This fix will be included in which release of rails? 5.2.3 (if planned) or 6.0.0? and when? please

**Comment by kamipo** [comment]
This fix is included in the next release 5.2.3 and 6.0.0.
The release schedule is uncertain. So if you need the fix now, please use 5-2-stable branch for now.
